### PR TITLE
feat(management): add broken link fixing cleanup rules

### PR DESCRIPTION
**Pull request by HussainTaj-arbisoft**
# What are the relevant tickets?

Relates to https://github.com/mitodl/hq/issues/1737

# Description (What does it do?)

Adds new markdown cleaning rules to help fix some of the broken links.

This PR includes multiple rules:

- `course_absolute_link` rule: fixes relative course (homepage) URLs.
    - For example, the course links on this page: https://ocw.mit.edu/courses/6-270-autonomous-robot-design-competition-january-iap-2005 
- `link_resolveuid` rule: fixes broken links that rely on the legacy `resolveuid` by finding their relevant WebsiteContent and replacing the links with `resource_link` shortcodes.
    - For example, the "back" link on this page: https://ocw.mit.edu/courses/15-071-the-analytics-edge-spring-2017/pages/an-introduction-to-analytics/assignment-1/
- `broken_metadata_link_fix` and `broken_markdown_link_fix`: these rules try to fix broken links by making an educated guess. 
    - For example:
       - Some links on https://ocw.mit.edu/courses/res-21g-01-kana-spring-2010 end with a `_index`.
       - The `video lectures` link on https://ocw.mit.edu/courses/2-993-special-topics-in-mechanical-engineering-the-art-and-science-of-boat-design-january-iap-2007/pages/assignments/ should refer to a video gallery but refers to a page instead.

## Sample Outputs

The following are outputs of the rules run on a production snapshot.

- [course_absolute_link](https://docs.google.com/spreadsheets/d/1wvGNzvdjctoZuYclP4p739HlxAMtup5jT-zzuWCfw90/edit?usp=sharing)
- [link_resolveuid](https://docs.google.com/spreadsheets/d/1KWclRZHcvj0fRAhbD69stuEEjhuX7h_OBcOhMMzN5AA/edit?usp=sharing)
- [broken_metadata_link_fix](https://docs.google.com/spreadsheets/d/1ZFXRObZd4gHrOWtMGFf0yL1U0naTbBgSAgf4Et1Hkws/edit?usp=sharing)
- [broken_markdown_link_fix](https://docs.google.com/spreadsheets/d/1RMFsEqq0U-L2Su5QQw0L2N5biLHPI83SqpEFMxVcIt4/edit?usp=sharing)

# How can this be tested?

## Prerequisites 

- A production or RC DB clone.

## Steps

1. Navigate to your local `ocw-studio` setup.
2. Checkout branch `hussaintaj/1737-broken-links`.
3. Start/Restart Studio.
4. Run the following set of commands.
    ```sh
    docker compose exec web ./manage.py markdown_cleanup course_absolute_link --skip-sync --out course_absolute_link.csv
    docker compose exec web ./manage.py markdown_cleanup link_resolveuid --skip-sync --out link_resolveuid.csv
    docker compose exec web ./manage.py markdown_cleanup broken_metadata_link_fix --skip-sync --out broken_metadata_link_fix.csv
    docker compose exec web ./manage.py markdown_cleanup broken_markdown_link_fix --skip-sync --out broken_markdown_link_fix.csv
    ```
5. Analyze the output files.
6. *Expect* changes to be predictable and valid.
7. Choose specific courses to apply the changes to and test builds with.
8. Run the following set of commands for specific courses. These will run a rule for a specific course, reset the courses sync states, populate the course's pipelines, and sync the course markdown to the backend. So please make sure your GitHub backend is set up.
    ```sh
    COURSE_ID=<course-id>
    docker compose exec web ./manage.py markdown_cleanup <rule> --skip-sync --commit --filter $COURSE_ID
    docker compose exec web ./manage.py reset_sync_states --skip_sync --filter $COURSE_ID && docker compose exec web ./manage.py backpopulate_pipelines --filter $COURSE_ID && docker compose exec web ./manage.py sync_website_to_backend --filter $COURSE_ID --delete
    ```
9. Publish your changes from the Studio UI.
10. *Expect* the build to succeed.
11. *Expect* the links to be fixed.

### CODE REVIEW

**Comment by pdpinch** [comment]
Can you say more about how this works and which broken legacy links it addresses?

**Comment by HussainTaj-arbisoft** [comment]
> Can you say more about how this works and which broken legacy links it addresses?

I've updated the PR description with a brief explanation. This PR follows the suggestion made in https://github.com/mitodl/hq/issues/1737#issuecomment-1771122605.

An example of the changes `link_resolveuid` rule will make can be found here: https://docs.google.com/spreadsheets/d/1AOjg331KXWomgGTM89HtwxK9AkswnrvCHYGY_-Ejy4s/edit?usp=sharing (this list can change as I test out things, I'll attach updated list when needed)

File: websites/management/commands/markdown_cleaning/rules/broken_link_fix.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+            NotImplementedError: When not implemented.
+        """
+        raise NotImplementedError
+
+    def _find_best_matching_content(
+        self, url: ParseResult, wc: WebsiteContent
+    ) -> WebsiteContent | None:
+        """
+        Use the last segment of the url path to find a matching
+        content in any of the content folders/paths.
```
Comment (ChristopherChudzicki) [inline websites/management/commands/markdown_cleaning/rules/broken_link_fix.py:113]: Here, too, I think some examples in the docstring would be helpful.

File: websites/management/commands/markdown_cleaning/rules/broken_link_fix.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+            replacement, notes = result.original_text, Notes(issue_type="Nothing to do")
+        else:
+            # Start making guesses.
+            found_wc = self._find_best_matching_content(url, wc)
+            notes = Notes(issue_type="Unknown link")
+
+            if found_wc:
+                replacement, fix_note = self.create_replacement(result, url, found_wc)
+                notes = Notes(
+                    issue_type="Unknown link",
```
Comment (ChristopherChudzicki) [inline websites/management/commands/markdown_cleaning/rules/broken_link_fix.py:207]:
```suggestion
            notes = Notes(issue_type="Found unique fuzzy match")
```
or something?  To distinguish from the Note just above.

File: websites/management/commands/markdown_cleaning/rules/broken_link_fix.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        "metadata.description",
+        "metadata.course_description",
+    ]
+
+    def create_replacement(
+        self, result: LinkParseResult, url: ParseResult, wc: WebsiteContent
+    ) -> (str, str):
+        """
+        Return a markdown link replacement text for `wc` (because
+        shortcodes are not supported in metadata yet.)
```
Comment (ChristopherChudzicki) [inline websites/management/commands/markdown_cleaning/rules/broken_link_fix.py:272]: This was true when I worked on markdown cleanup in March 2022, but it's not entirely true now: if metadata content is rendered with `renderString`, then shortcodes will be rendered.

https://github.com/gohugoio/hugo/issues/6703

I believe we generally use `renderString` to render metadata content, e.g., course description.

File: websites/management/commands/markdown_cleaning/rules/link_resolveuid.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                match_type="legacy website uid"
+            )
+
+        # text matches a content's title
+        contents = WebsiteContent.objects.filter(title=text, website=website)
+        if contents.count() == 1:
+            content = contents.first()
+            return ShortcodeTag.resource_link(
+                content.text_id, text, urlparse(result.link.destination).fragment
+            ).to_hugo(), Notes(match_type="content title")
```
Comment (ChristopherChudzicki) [inline websites/management/commands/markdown_cleaning/rules/link_resolveuid.py:97]: Seems reasonable 👍

**Review by ChristopherChudzicki — COMMENTED** [review]
This makes sense to me so far 👍 I have not actually tried running it yet.

By the way, in case it's helpful, my workflow when working on this stuff was often:
1. Run `markdonw_cleanup rule_name --out rule.csv`
2. Import the CSV into a database viewer for easy manipulation / QA. I use [TablePlus](https://tableplus.com/) for this.

File: websites/management/commands/markdown_cleaning/rules/broken_link_fix.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+        if `url_path` is a relative and valid course path.
+        """
+        try:
+            website_by_path = self.content_lookup.find_website_by_url_path(url_path)
+        except KeyError:
+            website_by_path = None
+
+        if website_by_path and website_by_path.unpublish_status is None:
+            # This is a course URL.
+            # These need to be absolute URLs to work correctly.
```
[the remark this answers is not in the document]
Comment (HussainTaj-arbisoft) [reply websites/management/commands/markdown_cleaning/rules/broken_link_fix.py:156]: This part is relevant to a specific issue. Some content has links to other courses (home pages) with a relative link, for example, `courses/1-050-solid-mechanics-fall-2004`.

These links need to be like `/courses/1-050-solid-mechanics-fall-2004`.

This rule is quite bloated, I'll take your advice and move this entire thing to a separate rule.

File: websites/management/commands/markdown_cleaning/rules/broken_link_fix.py (hunk 3 above)
Comment (HussainTaj-arbisoft) [reply websites/management/commands/markdown_cleaning/rules/broken_link_fix.py:272]: I see. That's awesome. Thanks for the info.

I will change the doc. Would you also like me to change the replacement text? I looked at Studio and don't think authors can edit shortcodes in metadata fields yet.

File: websites/management/commands/markdown_cleaning/utils.py [hunk 6]
```diff
@@ -25,8 +26,31 @@ def remove_prefix(string: str, prefix: str):
 def get_rootrelative_url_from_content(content: WebsiteContent):
     dirpath = remove_prefix(content.dirpath, "content/")
     filename = "" if content.filename == "_index" else content.filename
-    pieces = ["/courses", content.website.name, dirpath, filename]
-    return "/".join(p for p in pieces if p)
+    pieces = [content.website.url_path, dirpath, filename]
+    return "/" + "/".join(p for p in pieces if p)
```
Comment (HussainTaj-arbisoft) [inline websites/management/commands/markdown_cleaning/utils.py:30]: For most courses, `website.name` and `website.url_path` are the same. That is true for legacy courses. For newer courses, these two values can be different. And it is more appropriate to use `url_path`. So I've updated the utility accordingly.

File: websites/management/commands/markdown_cleaning/rules/broken_link_fix.py (hunk 3 above)
Comment (ChristopherChudzicki) [reply websites/management/commands/markdown_cleaning/rules/broken_link_fix.py]: >  I looked at Studio and don't think authors can edit shortcodes in metadata fields yet.

Good point.

I don't think you need to change the replacement text. Converting metadata links to shortcodes is something we should consider, but probably best handled separately.

**Review by ChristopherChudzicki — APPROVED** [review]
I did not try publishing the courses, but I went over the changes pretty carefully and they all look good!

I got:
```
rule                     ,changed links
broken_markdown_link_fix ,           23
broken_metadata_link_fix ,           50
course_absolute_link     ,           32
link_resolveuid          ,          170
```

All the changes look good, and the docstring examples are very helpful.

File: websites/management/commands/markdown_cleaning/rules/__init__.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+from importlib import import_module
+from inspect import isclass
+from pathlib import Path
+from pkgutil import iter_modules
+
+from ..cleaner import MarkdownCleanupRule  # noqa: TID252
```
[the remark this answers is not in the document]
Comment (HussainTaj-arbisoft) [reply websites/management/commands/markdown_cleaning/rules/__init__.py:6]: It was a personal preference. I think the absolute import is too long. 🤔 I'm not sure if everyone would agree. I'll change it.

```py
from websites.management.commands.markdown_cleaning.cleaner import MarkdownCleanupRule
```

**Pull request merged**